### ISSUE TITLE
Button should return number of clicks?

### ISSUE THREAD

**Issue by fonsp** [fonsp/PlutoUI.jl#38]
Right now the Button always returns the same: the label of the button (String). This is fine, you can use the button to "trigger reactivity" just by referencing the bound variable somewhere:

https://github.com/fonsp/PlutoUI.jl/blob/v0.6.1/src/Builtins.jl#L64-L86

But if you only want to do something _after_ the button is clicked, you need to write that logic yourself:

```julia
begin
    clicked_ref = Ref(false)
	@bind click Button("Do it!")
end
```

```julia
begin
	if clicked_ref[]
		something()
	end
    clicked_ref[] = true
end
```

Instead we could have the bound value be the number of times it was clicked:
```julia
@bind click Button("Do it!")
```

```julia
if click > 0
	something()
end
```

**Comment by fonsp** [comment]
@GiggleLiu

**Comment by fonsp** [comment]
Another option is to add a new element: `ToggleButton`, which looks like a button, but feels like a checkbox

**Comment by GiggleLiu** [comment]
Returning the counter is definitely better, besides providing more information, it also makes it clear that a click changes a value. It took me a while to understand the meaning of a button, and the answer is nothing :rofl: 

Returning a counter is not enough, a loophole of the above solution is
```julia
if click > 0
        other_vars
	something()
end
```
When `click > 0` and `other_vars` is updated, it can still run.

I posted a possible solution here
https://github.com/fonsp/PlutoUI.jl/issues/37#issuecomment-694432874

**Comment by fonsp** [comment]
But that's reactivity 😅 I'll comment more on the other issue

**Comment by GiggleLiu** [comment]
Yeah, I mean it requires people to understand the logic of Pluto. Is it possible to let button return a button object so that:

```julia
if button.signal()
    # do something
end
```

There is a field `_signal` in the button object, what `signal()` function do is,
* set `_signal = true` if the button is clicked,
* return the value of `_signal` when calling `signal()`,
* `signal()` function will reset the `_signal` field to false.

**Comment by dralletje** [comment]
```julia
onclick(button) do
	... 
end
```
:o

**Comment by fonsp** [comment]
https://github.com/fonsp/PlutoUI.jl/issues/37#issuecomment-695217352

**Comment by DhruvaSambrani** [comment]
But won't returning the number of times it is clicked generate a hidden state, namely the number of clicks? If someone wants an incremental counter, they can simply trigger reactivity in a cell which also updates a variable.

Or implement a Counter which can be both incremented and decreased. A button shouldn't return how many times it's clicked, just the fact that it is clicked.

**Comment by rapus95** [comment]
How about allowing to bind an already existing function name to the Button which then calls the function on click?
```julia
begin
  function onclick(text)
    dostuff()
    println("Button $text has been pressed")
  end
  @bind onclick Button("Do it!")
end
```

**Comment by fonsp** [comment]
@rapus95 It's a cool idea, but if you think about, it does not work at all with the way you write interactivity in Pluto. The cool thing about Pluto is that, instead of callbacks, we work with global variables as the way to make things interactive. In fact, it is not possible for a function to produce any kind of visual output, since the are no display side-effects in Pluto.

**Comment by rapus95** [comment]
@fonsp While that's true, there's no good way to barrier a cell from arbitrary evaluation on startup. We'd need an option to make a cell ONLY execute on the click of a button. I.e. there's no way to hold evaluation back until the click.

**Comment by fonsp** [comment]
```julia
@bind num_clicks Button()
```
```julia
if num_clicks > 0
	...
end
```

**Comment by rapus95** [comment]
And the counter is reset once the cell with the @bind is re-evaluated, right? Then that would be awesome. Is there an eta for it? :D

**Comment by fonsp** [comment]
I wrote the code for this as an example: 

https://plutocon2021-demos.netlify.app/fonsp%20%E2%80%94%20javascript%20inside%20pluto

So I suppose that it's already done :)

**Comment by rapus95** [comment]
> In fact, it is not possible for a function to produce any kind of visual output, since the are no display side-effects in Pluto.

That could individually be solved by adding some sort of `triggerupdate(:trigger)` function that allows for manually triggering the flow graph for certain variables. Though, I see that would make it necessary to wrap every cell into its own function and pass it a global state object 🤔 Probably not worth the effort.

Will the feature that the button references the click counter make it into Pluto? Arbitrarily breaking out of reactivity would also be very easy since it just requires an additional ref that holds a "counter at last time executed" number or any other relative number. I.e. triggering by comparison to a reference value.

**Comment by rapus95** [comment]
Together with https://github.com/fonsp/disorganised-mess/blob/f0f13eb4038604f0cccf12922cea6a4c635c6633/javascript.jl#L97-L128
(and div replaced by span in order to allow inline)
the following code
```julia
begin
	reactivitybreaker = Dict{Symbol, Int}()
	function needsupdate(x::Symbol, value)
		value != get(reactivitybreaker, x, 0)
	end
	function update!(x::Symbol, value; force=false)
		if force || needsupdate(x, value)
			reactivitybreaker[x] = value
			return true
		end
		return false
	end
	macro update!(x)
		return Expr(:call, :update!, QuoteNode(x), esc(x))
	end
end
```
allows for
```julia
@bind clicker ClickCounter("Click It!")
```
to be the only way to trigger the following print
```julia
if @update!(clicker)
  println(clicker)
end
```
